src/Provider.js

```diff
@@ -1,40 +1,35 @@
-import { Children, Component, createContext, createElement } from "react"
+import React from "react"
 import { shallowEqual } from "./utils/utils"
 
-export const MobXProviderContext = createContext({})
+export const MobXProviderContext = React.createContext({})
 
-export class Provider extends Component {
-    static contextType = MobXProviderContext
-    static displayName = "MobXProvider"
-
-    constructor(props, context) {
-        super(props, context)
-        this.state = {
-            ...context,
+export function Provider(props) {
+    const parentValue = React.useContext(MobXProviderContext)
+    const valueRef = React.useRef(null)
+    if (valueRef.current === null) {
+        valueRef.current = {
+            ...parentValue,
             ...grabStores(props)
         }
     }
+    const value = valueRef.current
 
-    render() {
-        return createElement(
-            MobXProviderContext.Provider,
-            { value: this.state },
-            this.props.children
-        )
-    }
-
-    static getDerivedStateFromProps(nextProps, prevState) {
-        if (process.env.NODE_ENV !== "production") {
-            const newStores = { ...prevState, ...grabStores(nextProps) } // spread in prevState for the context based stores
-            if (!shallowEqual(prevState, newStores))
-                throw new Error(
-                    "MobX Provider: The set of provided stores has changed. Please avoid changing stores as the change might not propagate to all children"
-                )
+    if (process.env.NODE_ENV !== "production") {
```

Comment:
I would perhaps move this inside the `useEffect` so it doesn't delay the rendering.

Reply:
I think it is good idea to throw as early as possible. Also this condition will delay the rendering in development mode only.